File: python/arrays-and-hashing/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        neighbors = [(0, 0), (0, 1), (0, -1), (1, 0), (1, 1), (1, -1), (-1, 0), (-1, 1), (-1, -1)]
        centers = [(1, 1), (1, 4), (1, 7), (4, 1), (4, 4), (4, 7), (7, 1), (7, 4), (7, 7)]

        # O(1) loop (9x9) to check by row
        for i in range(len(board)):
            temp = {}
            for j in range(len(board[0])):
                if board[i][j] != "." and board[i][j] in temp:
                    return False
                temp[board[i][j]] = 1

        # O(1) loop (9x9) to check by column
        for i in range(len(board[0])):
            temp = {}
            for j in range(len(board)):
                if board[j][i] != "." and board[j][i] in temp:
                    return False
                temp[board[j][i]] = 1

        # O(1) loop (9x9) to check for boxes
        for i in centers:
            temp = {}
            for j in neighbors:
                if board[i[0] + j[0]][i[1] + j[1]] != "." and board[i[0] + j[0]][i[1] + j[1]] in temp:
                    return False
                temp[board[i[0] + j[0]][i[1] + j[1]]] = 1

        return True

File: c++/arrays-and-hashing/valid-sudoku.cc
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        std::unordered_set<char> characters;

        std::vector<std::tuple<int, int>> centers = {std::make_tuple(1, 1), std::make_tuple(1, 4), std::make_tuple(1, 7), std::make_tuple(4, 1), std::make_tuple(4, 4), std::make_tuple(4, 7), std::make_tuple(7, 1), std::make_tuple(7, 4), std::make_tuple(7, 7)};
        std::vector<std::tuple<int, int>> boxes = {std::make_tuple(0, 1), std::make_tuple(0, 0), std::make_tuple(0, -1), std::make_tuple(1, 1), std::make_tuple(1, -1), std::make_tuple(1, 0), std::make_tuple(-1, 1), std::make_tuple(-1, 0), std::make_tuple(-1, -1)};

        for (int i = 0; i < board.size(); ++i) {
            for (int j = 0; j < board[0].size(); ++j) {
                if (board[i][j] != '.' && characters.contains(board[i][j])) {
                    return false;
                }
                characters.insert(board[i][j]);
            }
            characters.clear();
        }

        for (int i = 0; i < board[0].size(); ++i) {
            for (int j = 0; j < board.size(); ++j) {
                if (board[j][i] != '.' && characters.contains(board[j][i])) {
                    return false;
                }
                characters.insert(board[j][i]);
            }
            characters.clear();
        }

        for (auto [i, j] : centers) {
            for (auto [n_i, n_j] : boxes) {
                if (board[i+n_i][j+n_j] != '.' && characters.contains(board[i+n_i][j+n_j])) {
                    return false;
                }
                characters.insert(board[i+n_i][j+n_j]);
            }
            characters.clear();
        }

        return true;
    }
};
